text/0000-fused-iterator.md

```diff
@@ -0,0 +1,247 @@
+- Feature Name: fused
+- Start Date: 2016-04-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add a marker trait `FusedIterator` to `std::iter` and implement it on `Fuse<I>` and
+applicable iterators and adapters. By implementing `FusedIterator`, an iterator
+promises to behave as if `Iterator::fuse()` had been called on it (i.e. return
+`None` forever after returning `None` once). Then, specialize `Fuse<I>` to be a
+no-op iff `I` implements `FusedIterator`.
+
+# Motivation
+[motivation]: #motivation
+
+Iterators are allowed to return whatever they want after returning `None` once.
+However, assuming that an iterator continues to return `None` can make
+implementing some algorithms/adapters easier. Therefore, `Fused` and
+`Iterator::fuse` exist. Unfortunately, the `Fused` iterator adapter introduces a
+noticeable overhead. Furthermore, many iterators (most if not all iterators in
+std) already act as if they were fused (this is considered to be the "polite"
+behavior). Therefore, it would be nice to be able to pay the `Fused` overhead
+iff necessary.
```

Comment:
> Did you read an extra "not" into that sentence?

Oh, I indeed did.

That being said, `iff` is equivalent to logic-↔ and 

“nice to be able to pay the `Fused` overhead” → “necessary” ∧ 
“necessary” → “nice to be able to pay the `Fused` overhead” 

doesn’t really look right to me. /me shrugs.